src/internet_identity/src/vc_mvp.rs

```diff
@@ -0,0 +1,232 @@
+            rp_id_alias_credential: SignedIdAlias {
+                id_alias: id_alias_principal,
+                id_dapp: id_rp,
+                credential_jws: rp_jws,
+            },
+            issuer_id_alias_credential: SignedIdAlias {
+                id_alias: id_alias_principal,
+                id_dapp: id_issuer,
+                credential_jws: issuer_jws,
+            },
+        })
+    })
+}
+
+fn get_signature(
+    cert_assets: &CertifiedAssets,
+    sigs: &SignatureMap,
+    seed: Hash,
+    msg_hash: Hash,
+) -> Option<Vec<u8>> {
+    let certificate = data_certificate().unwrap_or_else(|| {
+        trap("data certificate is only available in query calls");
+    });
+    let witness = sigs.witness(hash::hash_bytes(seed), msg_hash)?;
+
+    let witness_hash = witness.reconstruct();
+    let root_hash = sigs.root_hash();
+    if witness_hash != root_hash {
+        trap(&format!(
+            "internal error: signature map computed an invalid hash tree, witness hash is {}, root hash is {}",
+            hex::encode(witness_hash),
+            hex::encode(root_hash)
+        ));
+    }
+
+    let tree = ic_certified_map::fork(
+        HashTree::Pruned(cert_assets.root_hash()),
+        ic_certified_map::labeled(LABEL_SIG, witness),
+    );
+
+    #[derive(Serialize)]
+    struct Sig<'a> {
+        certificate: ByteBuf,
+        tree: HashTree<'a>,
+    }
+
+    let sig = Sig {
+        certificate: ByteBuf::from(certificate),
+        tree,
+    };
+
+    let mut cbor = serde_cbor::ser::Serializer::new(Vec::new());
+    cbor.self_describe().unwrap();
+    sig.serialize(&mut cbor).unwrap();
+    Some(cbor.into_inner())
+}
+
+fn add_signature(sigs: &mut SignatureMap, msg_hash: Hash, seed: Hash) {
+    let expires_at = time().saturating_add(SIGNATURE_EXPIRATION_PERIOD_NS);
+    sigs.put(hash::hash_bytes(seed), msg_hash, expires_at);
+}
+
+fn calculate_id_alias_seed(identity_number: AnchorNumber, dapps: &InvolvedDapps) -> Hash {
+    let salt = state::salt();
+
+    let mut blob: Vec<u8> = vec![];
+    blob.push(salt.len() as u8);
+    blob.extend_from_slice(&salt);
+
+    let identity_number_str = identity_number.to_string();
+    let identity_number_blob = identity_number_str.bytes();
+    blob.push(identity_number_blob.len() as u8);
+    blob.extend(identity_number_blob);
+
+    blob.push(dapps.relying_party.bytes().len() as u8);
+    blob.extend(dapps.relying_party.bytes());
+
+    blob.push(dapps.issuer.bytes().len() as u8);
+    blob.extend(dapps.issuer.bytes());
+
+    hash::hash_bytes(blob)
+}
+
+fn id_alias_credential(alias_tuple: &AliasTuple) -> Credential {
+    let subject: Subject = Subject::from_json_value(json!({
+        "id": did_for_principal(alias_tuple.id_dapp),
+        "has_id_alias": did_for_principal(alias_tuple.id_alias),
+    }))
+    .expect("internal: failed building id_alias subject");
+
+    let credential: Credential = CredentialBuilder::default()
```

Comment:
Maybe it would be prudent to put an `exp` claim inside as well? Otherwise we will break the semantics of older tokens if we ever change how the seed is calculated.
And it is also useful to emphasize the ephemeral nature of these credentials.

Reply:
Good idea, done.